sw/device/tests/sram_ctrl_execution_test.c

```diff
@@ -0,0 +1,94 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include <stdbool.h>
+#include <stdint.h>
+
+#include "sw/device/examples/sram_program/sram_program.h"
+#include "sw/device/lib/base/macros.h"
+#include "sw/device/lib/dif/dif_sram_ctrl.h"
+#include "sw/device/lib/runtime/ibex.h"
+#include "sw/device/lib/runtime/log.h"
+#include "sw/device/lib/testing/test_framework/check.h"
+#include "sw/device/lib/testing/test_framework/ottf_isrs.h"
+#include "sw/device/lib/testing/test_framework/ottf_macros.h"
+#include "sw/device/lib/testing/test_framework/ottf_main.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+
+OTTF_DEFINE_TEST_CONFIG();
+
+static dif_sram_ctrl_t sram_ctrl;
+
+enum {
+  kSramStart = TOP_EARLGREY_SRAM_CTRL_MAIN_RAM_BASE_ADDR,
+  kSramEnd = TOP_EARLGREY_SRAM_CTRL_MAIN_RAM_BASE_ADDR +
+             TOP_EARLGREY_SRAM_CTRL_MAIN_RAM_SIZE_BYTES,
+  kSramRetStart = TOP_EARLGREY_SRAM_CTRL_RET_AON_RAM_BASE_ADDR,
+};
+
+OT_SECTION(".data.sram_function_test")
+OT_NOINLINE()
+static void sram_function_test(void) {
+  uint32_t pc = 0;
+  asm("auipc %[pc], 0;" : [pc] "=r"(pc));
+  LOG_INFO("PC: %p, SRAM: [%p, %p)", pc, kSramStart, kSramEnd);
+  CHECK(pc >= kSramStart && pc < kSramEnd, "PC is outside the main SRAM");
+}
+
+// Redeclaration of the addressable symbol in `sram_ret_neg_test()` to be used
+// in `ottf_exception_handler()`.
+extern const char kSramRetNegTestReturn[];
+
+void ottf_exception_handler(void) {
+  // The frame address is the address of the stack location that holds the
+  // `mepc`, since the OTTF exception handler entry code saves the `mepc` to
+  // the top of the stack before transferring control flow to the exception
+  // handler function (which is overridden here). See the `handler_exception`
+  // subroutine in `sw/device/lib/testing/testing/ottf_isrs.S` for more details.
+  uintptr_t mepc_stack_addr = (uintptr_t)OT_FRAME_ADDR();
+  ottf_exc_id_t exception_id = ibex_mcause_read();
+  switch (exception_id) {
+    case kInstrAccessFault:
+      LOG_INFO("Detected instruction access fault");
+      *(uintptr_t *)mepc_stack_addr = (uintptr_t)kSramRetNegTestReturn;
```

Comment:
This line caught my eye! Does it make sense to reinterpret `mepc_stack_addr` as a `uintptr_t*`, or should it just be declared as that type to begin with?

Reply:
Good catch! I just kept the types of the original code, changed type to `uintptr_t*`